TestLiveSigstoreVerifier/with_2/3_verified_attestations is flaky

### Describe the bug

TestLiveSigstoreVerifier/with_2/3_verified_attestations occasionally fails

### Affected version

PRs to trunk

### Steps to reproduce the behavior

1. Type this '...'
2. View the output '....'
3. See error

### Expected vs actual behavior

Expected: pass

Actual:

https://github.com/cli/cli/actions/workflows/go.yml?query=is%3Afailure

#### windows-latest
https://github.com/cli/cli/actions/runs/13184622557/job/36803634377#step:5:59
```

--- FAIL: TestLiveSigstoreVerifier (3.12s)
    --- FAIL: TestLiveSigstoreVerifier/with_2/3_verified_attestations (1.12s)
        sigstore_integration_test.go:79: 
            	Error Trace:	D:/a/cli/cli/pkg/cmd/attestation/verification/sigstore_integration_test.go:79
            	Error:      	"[0xc00004b0f0]" should have 2 item(s), but has 1
            	Test:       	TestLiveSigstoreVerifier/with_2/3_verified_attestations
FAIL
FAIL	github.com/cli/cli/v2/pkg/cmd/attestation/verification	3.468s
```

https://github.com/cli/cli/actions/runs/13202858867/job/36858853456#step:5:61

```
--- FAIL: TestVerifyAttestations (1.78s)
    --- FAIL: TestVerifyAttestations/passes_verification_with_2/3_attestations_passing_Sigstore_verification (0.37s)
        attestation_integration_test.go:65: 
            	Error Trace:	D:/a/cli/cli/pkg/cmd/attestation/verify/attestation_integration_test.go:65
            	Error:      	"[0xc000875820]" should have 2 item(s), but has 1
            	Test:       	TestVerifyAttestations/passes_verification_with_2/3_attestations_passing_Sigstore_verification
FAIL
FAIL	github.com/cli/cli/v2/pkg/cmd/attestation/verify	12.350s
```

#### macos-latest
https://github.com/cli/cli/actions/runs/12916504935/job/36020835039#step:5:60
```
--- FAIL: TestVerifyAttestations (1.45s)
    --- FAIL: TestVerifyAttestations/all_attestations_pass_verification (0.37s)
        attestation_integration_test.go:53: 
            	Error Trace:	/Users/runner/work/cli/cli/pkg/cmd/attestation/verify/attestation_integration_test.go:53
            	Error:      	"[0xc000409e[60](https://github.com/cli/cli/actions/runs/12916504935/job/36020835039#step:5:61)]" should have 2 item(s), but has 1
            	Test:       	TestVerifyAttestations/all_attestations_pass_verification
FAIL
```

#### ubuntu-latest
https://github.com/cli/cli/actions/runs/12680772503/job/35343196076#step:5:64
```
--- FAIL: TestVerifyAttestations (1.32s)
    --- FAIL: TestVerifyAttestations/passes_verification_with_2/3_attestations_passing_Sigstore_verification (0.26s)
        attestation_integration_test.go:[65](https://github.com/cli/cli/actions/runs/12680772503/job/35343196076#step:5:66): 
            	Error Trace:	/home/runner/work/cli/cli/pkg/cmd/attestation/verify/attestation_integration_test.go:65
            	Error:      	"[0xc0009[67](https://github.com/cli/cli/actions/runs/12680772503/job/35343196076#step:5:68)ae0]" should have 2 item(s), but has 1
            	Test:       	TestVerifyAttestations/passes_verification_with_2/3_attestations_passing_Sigstore_verification
FAIL
FAIL	github.com/cli/cli/v2/pkg/cmd/attestation/verify	6.048s
```

https://github.com/cli/cli/actions/runs/12659904205/job/35279973709#step:5:64
```
--- FAIL: TestVerifyAttestations (1.34s)
    --- FAIL: TestVerifyAttestations/all_attestations_pass_verification (0.45s)
        attestation_integration_test.go:53: 
            	Error Trace:	/home/runner/work/cli/cli/pkg/cmd/attestation/verify/attestation_integration_test.go:53
            	Error:      	"[0xc0008511f0]" should have 2 item(s), but has 1
            	Test:       	TestVerifyAttestations/all_attestations_pass_verification
FAIL
FAIL	github.com/cli/cli/v2/pkg/cmd/attestation/verify	6.936s
```

### Logs

Paste the activity from your command line. Redact if needed.

<!-- Note: Set `GH_DEBUG=true` for verbose logs or `GH_DEBUG=api` for verbose logs with HTTP traffic details. -->

Getting the same error for Windows:

```
--- FAIL: TestVerifyAttestations (1.43s)
    --- FAIL: TestVerifyAttestations/all_attestations_pass_verification (0.41s)
        attestation_integration_test.go:53: 
            	Error Trace:	D:/a/cli/cli/pkg/cmd/attestation/verify/attestation_integration_test.go:53
            	Error:      	"[0xc00090fdf0]" should have 2 item(s), but has 1
            	Test:       	TestVerifyAttestations/all_attestations_pass_verification
```

Workflow: https://github.com/cli/cli/actions/runs/13251671571/job/36990690943?pr=10334#step:5:61

---

Here's another one:

```
--- FAIL: TestVerifyAttestations (1.77s)
    --- FAIL: TestVerifyAttestations/passes_verification_with_2/3_attestations_passing_Sigstore_verification (0.39s)
        attestation_integration_test.go:65: 
            	Error Trace:	D:/a/cli/cli/pkg/cmd/attestation/verify/attestation_integration_test.go:65
            	Error:      	"[0xc000391ef0]" should have 2 item(s), but has 1
            	Test:       	TestVerifyAttestations/passes_verification_with_2/3_attestations_passing_Sigstore_verification
```

Workflow: https://github.com/cli/cli/actions/runs/13258224226/job/37008805848?pr=10413#step:5:61

---

**2025-02-12 (Ubuntu)**

```
--- FAIL: TestVerifyIntegration (18.63s)
    --- FAIL: TestVerifyIntegration/with_invalid_owner (15.17s)
        verify_integration_test.go:76: 
            	Error Trace:	/home/runner/work/cli/cli/pkg/cmd/attestation/verify/verify_integration_test.go:76
            	Error:      	Error "failed to find recognized issuer from bundle content: failed to create TUF client: tuf refresh failed: context deadline exceeded (Client.Timeout or context cancellation while reading body)" does not contain "expected SourceRepositoryOwnerURI to be https://github.com/fakeowner, got https://github.com/sigstore"
            	Test:       	TestVerifyIntegration/with_invalid_owner
```

Workflow: https://github.com/cli/cli/actions/runs/13288202067/job/37102016536#step:5:64

---

**2025-02-13 (macOS)**

```
WARNING: DATA RACE
Read at 0x00c000440ae0 by goroutine 72:
  github.com/cli/cli/v2/pkg/cmd/attestation/api.(*failAfterNCallsHttpClient).Get()
      /Users/runner/work/cli/cli/pkg/cmd/attestation/api/mock_httpClient_test.go:69 +0x68
  github.com/cli/cli/v2/pkg/cmd/attestation/api.(*LiveClient).getBundle.func1()
      /Users/runner/work/cli/cli/pkg/cmd/attestation/api/client.go:194 +0x78
  github.com/cenkalti/backoff/v4.RetryNotifyWithTimer.Operation.withEmptyData.func1()
      /Users/runner/go/pkg/mod/github.com/cenkalti/backoff/v4@v4.3.0/retry.go:18 +0x30
  github.com/cenkalti/backoff/v4.doRetryNotify[go.shape.struct {}]()
      /Users/runner/go/pkg/mod/github.com/cenkalti/backoff/v4@v4.3.0/retry.go:88 +0x15c
  github.com/cenkalti/backoff/v4.RetryNotifyWithTimer()
      /Users/runner/go/pkg/mod/github.com/cenkalti/backoff/v4@v4.3.0/retry.go:61 +0x80
  github.com/cenkalti/backoff/v4.RetryNotify()
      /Users/runner/go/pkg/mod/github.com/cenkalti/backoff/v4@v4.3.0/retry.go:49 +0x1b0
  github.com/cenkalti/backoff/v4.Retry()
      /Users/runner/go/pkg/mod/github.com/cenkalti/backoff/v4@v4.3.0/retry.go:38 +0x190
  github.com/cli/cli/v2/pkg/cmd/attestation/api.(*LiveClient).getBundle()
      /Users/runner/work/cli/cli/pkg/cmd/attestation/api/client.go:193 +0x138
  github.com/cli/cli/v2/pkg/cmd/attestation/api.(*LiveClient).fetchBundleFromAttestations.func1()
      /Users/runner/work/cli/cli/pkg/cmd/attestation/api/client.go:169 +0x1f0
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /Users/runner/go/pkg/mod/golang.org/x/sync@v0.10.0/errgroup/errgroup.go:78 +0x7c

Previous write at 0x00c000440ae0 by goroutine 73:
  github.com/cli/cli/v2/pkg/cmd/attestation/api.(*failAfterNCallsHttpClient).Get()
      /Users/runner/work/cli/cli/pkg/cmd/attestation/api/mock_httpClient_test.go:69 +0x7c
  github.com/cli/cli/v2/pkg/cmd/attestation/api.(*LiveClient).getBundle.func1()
      /Users/runner/work/cli/cli/pkg/cmd/attestation/api/client.go:194 +0x78
  github.com/cenkalti/backoff/v4.RetryNotifyWithTimer.Operation.withEmptyData.func1()
      /Users/runner/go/pkg/mod/github.com/cenkalti/backoff/v4@v4.3.0/retry.go:18 +0x30
  github.com/cenkalti/backoff/v4.doRetryNotify[go.shape.struct {}]()
      /Users/runner/go/pkg/mod/github.com/cenkalti/backoff/v4@v4.3.0/retry.go:88 +0x15c
  github.com/cenkalti/backoff/v4.RetryNotifyWithTimer()
      /Users/runner/go/pkg/mod/github.com/cenkalti/backoff/v4@v4.3.0/retry.go:61 +0x80
  github.com/cenkalti/backoff/v4.RetryNotify()
      /Users/runner/go/pkg/mod/github.com/cenkalti/backoff/v4@v4.3.0/retry.go:49 +0x1b0
  github.com/cenkalti/backoff/v4.Retry()
      /Users/runner/go/pkg/mod/github.com/cenkalti/backoff/v4@v4.3.0/retry.go:38 +0x190
  github.com/cli/cli/v2/pkg/cmd/attestation/api.(*LiveClient).getBundle()
      /Users/runner/work/cli/cli/pkg/cmd/attestation/api/client.go:193 +0x138
  github.com/cli/cli/v2/pkg/cmd/attestation/api.(*LiveClient).fetchBundleFromAttestations.func1()
      /Users/runner/work/cli/cli/pkg/cmd/attestation/api/client.go:169 +0x1f0
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /Users/runner/go/pkg/mod/golang.org/x/sync@v0.10.0/errgroup/errgroup.go:78 +0x7c

Goroutine 72 (running) created at:
  golang.org/x/sync/errgroup.(*Group).Go()
      /Users/runner/go/pkg/mod/golang.org/x/sync@v0.10.0/errgroup/errgroup.go:75 +0x10c
  github.com/cli/cli/v2/pkg/cmd/attestation/api.(*LiveClient).fetchBundleFromAttestations()
      /Users/runner/work/cli/cli/pkg/cmd/attestation/api/client.go:1[54](https://github.com/cli/cli/actions/runs/13301543234/job/37143755142?pr=10430#step:5:55) +0xb4
  github.com/cli/cli/v2/pkg/cmd/attestation/api.TestFetchBundleFromAttestations_FailOnTheSecondAttestation()
      /Users/runner/work/cli/cli/pkg/cmd/attestation/api/client_test.go:215 +0x2e0
  testing.tRunner()
      /Users/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.darwin-arm64/src/testing/testing.go:1690 +0x184
  testing.(*T).Run.gowrap1()
      /Users/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.darwin-arm64/src/testing/testing.go:1743 +0x40

Goroutine 73 (running) created at:
  golang.org/x/sync/errgroup.(*Group).Go()
      /Users/runner/go/pkg/mod/golang.org/x/sync@v0.10.0/errgroup/errgroup.go:75 +0x10c
  github.com/cli/cli/v2/pkg/cmd/attestation/api.(*LiveClient).fetchBundleFromAttestations()
      /Users/runner/work/cli/cli/pkg/cmd/attestation/api/client.go:154 +0xb4
  github.com/cli/cli/v2/pkg/cmd/attestation/api.TestFetchBundleFromAttestations_FailOnTheSecondAttestation()
      /Users/runner/work/cli/cli/pkg/cmd/attestation/api/client_test.go:215 +0x2e0
  testing.tRunner()
      /Users/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.darwin-arm64/src/testing/testing.go:1690 +0x184
  testing.(*T).Run.gowrap1()
      /Users/runner/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.darwin-arm64/src/testing/testing.go:1743 +0x40
==================
--- FAIL: TestFetchBundleFromAttestations_FailOnTheSecondAttestation (0.[63](https://github.com/cli/cli/actions/runs/13301543234/job/37143755142?pr=10430#step:5:64)s)
    testing.go:1399: race detected during execution of test
FAIL
```

Workflow: https://github.com/cli/cli/actions/runs/13301543234/job/37143755142?pr=10430#step:5:54

@malancas: It's still failing:

```
--- FAIL: TestLiveSigstoreVerifier (2.24s)
    --- FAIL: TestLiveSigstoreVerifier/with_2/3_verified_attestations (0.42s)
        sigstore_integration_test.go:79: 
            	Error Trace:	D:/a/cli/cli/pkg/cmd/attestation/verification/sigstore_integration_test.go:79
            	Error:      	"[0xc0002c0780]" should have 2 item(s), but has 1
            	Test:       	TestLiveSigstoreVerifier/with_2/3_verified_attestations
FAIL
```

Workflow: https://github.com/iamazeem/cli/actions/runs/13334014863/job/37244936343#step:5:60

---

**2025-02-16 (Windows)**

```
--- FAIL: TestLiveSigstoreVerifier (1.98s)
    --- FAIL: TestLiveSigstoreVerifier/with_2/3_verified_attestations (0.34s)
        sigstore_integration_test.go:79: 
            	Error Trace:	D:/a/cli/cli/pkg/cmd/attestation/verification/sigstore_integration_test.go:79
            	Error:      	"[0xc0004b3430]" should have 2 item(s), but has 1
            	Test:       	TestLiveSigstoreVerifier/with_2/3_verified_attestations
```

Workflow: https://github.com/cli/cli/actions/runs/13354330624/job/37294953903?pr=10456#step:5:59

---

**2025-02-18 (Windows)**

```
--- FAIL: TestLiveSigstoreVerifier (3.02s)
    --- FAIL: TestLiveSigstoreVerifier/with_2/3_verified_attestations (0.48s)
        sigstore_integration_test.go:79: 
            	Error Trace:	D:/a/cli/cli/pkg/cmd/attestation/verification/sigstore_integration_test.go:79
            	Error:      	"[0xc00018b8c0]" should have 2 item(s), but has 1
            	Test:       	TestLiveSigstoreVerifier/with_2/3_verified_attestations
FAIL
FAIL	github.com/cli/cli/v2/pkg/cmd/attestation/verification	3.302s
?   	github.com/cli/cli/v2/pkg/cmd/auth/shared/contract	[no test files]
--- FAIL: TestVerifyAttestations (1.80s)
    --- FAIL: TestVerifyAttestations/all_attestations_pass_verification (0.61s)
        attestation_integration_test.go:53: 
            	Error Trace:	D:/a/cli/cli/pkg/cmd/attestation/verify/attestation_integration_test.go:53
            	Error:      	"[0xc0009265[60](https://github.com/cli/cli/actions/runs/13383511075/job/37376071790?pr=10456#step:5:61)]" should have 2 item(s), but has 1
            	Test:       	TestVerifyAttestations/all_attestations_pass_verification
FAIL
```

Workflow: https://github.com/cli/cli/actions/runs/13383511075/job/37376071790?pr=10456#step:5:59

**UPDATE**

Running tests locally on **Windows 10** also reproduces this.
However, it's a bit random but overall percentage is pretty high i.e. ~80%.

Command:

```
go test -tags=integration .\pkg\cmd\attestation\... -count 2 -failfast
```

Errors:

```
--- FAIL: TestVerifyAttestations (1.13s)
    --- FAIL: TestVerifyAttestations/all_attestations_pass_verification (1.12s)
        attestation_integration_test.go:53:
                Error Trace:    C:/cli-trunk/pkg/cmd/attestation/verify/attestation_integration_test.go:53
                Error:          "[0xc000719230]" should have 2 item(s), but has 1
                Test:           TestVerifyAttestations/all_attestations_pass_verification
```

```
--- FAIL: TestLiveSigstoreVerifier (3.09s)
    sigstore_integration_test.go:62:
                Error Trace:    C:/cli-trunk/pkg/cmd/attestation/verification/sigstore_integration_test.go:62
                Error:          Not equal:
                                expected: 2
                                actual  : 1
                Test:           TestLiveSigstoreVerifier
                Messages:       test case: with valid artifact and JSON lines file containing multiple Sigstore bundles
```

```
--- FAIL: TestVerifyAttestations (3.72s)
    --- FAIL: TestVerifyAttestations/passes_verification_with_2/3_attestations_passing_Sigstore_verification (1.42s)
        attestation_integration_test.go:65:
                Error Trace:    C:/cli-trunk/pkg/cmd/attestation/verify/attestation_integration_test.go:65
                Error:          "[0xc000781d30]" should have 2 item(s), but has 1
                Test:           TestVerifyAttestations/passes_verification_with_2/3_attestations_passing_Sigstore_verification
```

```
--- FAIL: TestVerifyAttestations (1.85s)
    --- FAIL: TestVerifyAttestations/passes_verification_with_2/3_attestations_passing_Sigstore_verification (0.18s)
        attestation_integration_test.go:63:
                Error Trace:    C:/cli-trunk/pkg/cmd/attestation/verify/attestation_integration_test.go:63
                Error:          Received unexpected error:
                                failed to get bundle issuer: unsupported bundle version: application/vnd.dev.sigstore.bundle+json;version=0.1
                Test:           TestVerifyAttestations/passes_verification_with_2/3_attestations_passing_Sigstore_verification
```

```
--- FAIL: TestLiveSigstoreVerifier (3.61s)
    --- FAIL: TestLiveSigstoreVerifier/with_2/3_verified_attestations (0.82s)
        sigstore_integration_test.go:79:
                Error Trace:    C:/cli-trunk/pkg/cmd/attestation/verification/sigstore_integration_test.go:79
                Error:          "[0xc000293310]" should have 2 item(s), but has 1
                Test:           TestLiveSigstoreVerifier/with_2/3_verified_attestations
```

---

Ran similar tests on **Ubuntu 22.04 LTS** with `-race` and `-count 100` and got this error trace:

```
$ go test -tags=integration ./pkg/cmd/attestation/verification/... -race -count 100 -failfast
panic: test timed out after 10m0s
	running tests:
		TestLiveSigstoreVerifier (2s)

goroutine 6324 [running]:
testing.(*M).startAlarm.func1()
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/testing/testing.go:2373 +0x265
created by time.goFunc
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/time/sleep.go:215 +0x45

goroutine 1 [chan receive]:
testing.(*T).Run(0xc0003fd1e0, {0xfbac6d, 0x18}, 0xffd168)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/testing/testing.go:1751 +0x851
testing.runTests.func1(0xc0003fd1e0)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/testing/testing.go:2168 +0x86
testing.tRunner(0xc0003fd1e0, 0xc00031bae0)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/testing/testing.go:1690 +0x227
testing.runTests(0xc0001aa2b8, {0x15abaa0, 0x8, 0x8}, {0x7f56d5faa108?, 0x40?, 0x15b86a0?})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/testing/testing.go:2166 +0x8bf
testing.(*M).Run(0xc0001be000)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/testing/testing.go:2034 +0xf18
main.main()
	_testmain.go:59 +0x165

goroutine 62 [IO wait]:
internal/poll.runtime_pollWait(0x7f56d57c6568, 0x72)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/runtime/netpoll.go:351 +0x85
internal/poll.(*pollDesc).wait(0xc00025c920, 0x72, 0x0)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/internal/poll/fd_poll_runtime.go:84 +0xb1
internal/poll.(*pollDesc).waitRead(...)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Read(0xc00025c900, {0xc000176a80, 0x2a80, 0x2a80})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/internal/poll/fd_unix.go:165 +0x466
net.(*netFD).Read(0xc00025c900, {0xc000176a80, 0x2a80, 0x2a80})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/fd_posix.go:55 +0x4b
net.(*conn).Read(0xc000196820, {0xc000176a80, 0x2a80, 0x2a80})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/net.go:189 +0xad
crypto/tls.(*atLeastReader).Read(0xc000362090, {0xc000176a80, 0x2a80, 0x2a80})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/crypto/tls/conn.go:809 +0x89
bytes.(*Buffer).ReadFrom(0xc0003fbeb8, {0x10e2fa0, 0xc000362090})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/bytes/buffer.go:211 +0x113
crypto/tls.(*Conn).readFromUntil(0xc0003fbc08, {0x10e2b20, 0xc000196820}, 0x5)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/crypto/tls/conn.go:831 +0x1da
crypto/tls.(*Conn).readRecordOrCCS(0xc0003fbc08, 0x0)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/crypto/tls/conn.go:629 +0x365
crypto/tls.(*Conn).readRecord(...)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/crypto/tls/conn.go:591
crypto/tls.(*Conn).Read(0xc0003fbc08, {0xc00057f000, 0x1000, 0x0?})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/crypto/tls/conn.go:1385 +0x2dd
bufio.(*Reader).Read(0xc00027f440, {0xc000237ee0, 0x9, 0x9})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/bufio/bufio.go:241 +0x4bf
io.ReadAtLeast({0x10e1360, 0xc00027f440}, {0xc000237ee0, 0x9, 0x9}, 0x9)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/io/io.go:335 +0xd0
io.ReadFull(...)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/io/io.go:354
net/http.http2readFrameHeader({0xc000237ee0, 0x9, 0x9}, {0x10e1360, 0xc00027f440})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/h2_bundle.go:1642 +0x9a
net/http.(*http2Framer).ReadFrame(0xc000237ea0)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/h2_bundle.go:1909 +0xf5
net/http.(*http2clientConnReadLoop).run(0xc000317fa8)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/h2_bundle.go:9496 +0x191
net/http.(*http2ClientConn).readLoop(0xc0001c5980)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/h2_bundle.go:9392 +0x175
created by net/http.(*http2Transport).newClientConn in goroutine 61
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/h2_bundle.go:8006 +0x1e2b

goroutine 186 [IO wait]:
internal/poll.runtime_pollWait(0x7f56d57c6450, 0x72)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/runtime/netpoll.go:351 +0x85
internal/poll.(*pollDesc).wait(0xc00025d420, 0x72, 0x0)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/internal/poll/fd_poll_runtime.go:84 +0xb1
internal/poll.(*pollDesc).waitRead(...)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Read(0xc00025d400, {0xc00039c600, 0x1300, 0x1300})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/internal/poll/fd_unix.go:165 +0x466
net.(*netFD).Read(0xc00025d400, {0xc00039c600, 0x1300, 0x1300})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/fd_posix.go:55 +0x4b
net.(*conn).Read(0xc0006a0010, {0xc00039c600, 0x1300, 0x1300})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/net.go:189 +0xad
crypto/tls.(*atLeastReader).Read(0xc000352030, {0xc00039c600, 0x1300, 0x1300})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/crypto/tls/conn.go:809 +0x89
bytes.(*Buffer).ReadFrom(0xc0000050b8, {0x10e2fa0, 0xc000352030})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/bytes/buffer.go:211 +0x113
crypto/tls.(*Conn).readFromUntil(0xc000004e08, {0x10e2b20, 0xc0006a0010}, 0x5)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/crypto/tls/conn.go:831 +0x1da
crypto/tls.(*Conn).readRecordOrCCS(0xc000004e08, 0x0)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/crypto/tls/conn.go:629 +0x365
crypto/tls.(*Conn).readRecord(...)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/crypto/tls/conn.go:591
crypto/tls.(*Conn).Read(0xc000004e08, {0xc00001b000, 0x1000, 0xc000319ac8?})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/crypto/tls/conn.go:1385 +0x2dd
bufio.(*Reader).Read(0xc0000ad740, {0xc0002362e0, 0x9, 0x9})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/bufio/bufio.go:241 +0x4bf
io.ReadAtLeast({0x10e1360, 0xc0000ad740}, {0xc0002362e0, 0x9, 0x9}, 0x9)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/io/io.go:335 +0xd0
io.ReadFull(...)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/io/io.go:354
net/http.http2readFrameHeader({0xc0002362e0, 0x9, 0x9}, {0x10e1360, 0xc0000ad740})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/h2_bundle.go:1642 +0x9a
net/http.(*http2Framer).ReadFrame(0xc0002362a0)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/h2_bundle.go:1909 +0xf5
net/http.(*http2clientConnReadLoop).run(0xc000319fa8)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/h2_bundle.go:9496 +0x191
net/http.(*http2ClientConn).readLoop(0xc00055a000)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/h2_bundle.go:9392 +0x175
created by net/http.(*http2Transport).newClientConn in goroutine 185
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/h2_bundle.go:8006 +0x1e2b

goroutine 6319 [select]:
net/http.(*http2ClientConn).roundTrip(0xc0001c5980, 0xc0004da3c0, 0x0)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/h2_bundle.go:8520 +0xc0f
net/http.(*http2ClientConn).RoundTrip(...)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/h2_bundle.go:8417
net/http.(*http2Transport).RoundTripOpt(0xc0000c06e0, 0xc0004da3c0, {0x74?})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/h2_bundle.go:7739 +0x64e
net/http.(*http2Transport).RoundTrip(...)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/h2_bundle.go:7697
net/http.http2noDialH2RoundTripper.RoundTrip({0x15acae0?}, 0xc0004da3c0)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/h2_bundle.go:10371 +0x38
net/http.(*Transport).roundTrip(0x15acae0, 0xc0004da3c0)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/transport.go:563 +0x637
net/http.(*Transport).RoundTrip(0x15acae0, 0xc0004da3c0)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/roundtrip.go:30 +0x34
net/http.send(0xc0004da280, {0x10e2000, 0x15acae0}, {0x0?, 0x0?, 0x15b86a0?})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/client.go:259 +0x9a7
net/http.(*Client).send(0xc00030d320, 0xc0004da280, {0x0?, 0x44aea5?, 0x15b86a0?})
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/client.go:180 +0x14f
net/http.(*Client).do(0xc00030d320, 0xc0004da280)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/client.go:731 +0x13b2
net/http.(*Client).Do(...)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/client.go:590
github.com/theupdateframework/go-tuf/v2/metadata/fetcher.(*DefaultFetcher).DownloadFile(0xc00025eb00, {0xc0003260f0, 0x30}, 0x4000, 0x37e11d600)
	/home/azeem/go/pkg/mod/github.com/theupdateframework/go-tuf/v2@v2.0.2/metadata/fetcher/fetcher.go:57 +0x2b0
github.com/theupdateframework/go-tuf/v2/metadata/updater.(*Updater).downloadMetadata(0xc00025e600, {0xfafaba, 0x9}, 0x4000, {0x0, 0x0})
	/home/azeem/go/pkg/mod/github.com/theupdateframework/go-tuf/v2@v2.0.2/metadata/updater/updater.go:640 +0x34b
github.com/theupdateframework/go-tuf/v2/metadata/updater.(*Updater).loadTimestamp(0xc00025e600)
	/home/azeem/go/pkg/mod/github.com/theupdateframework/go-tuf/v2@v2.0.2/metadata/updater/updater.go:325 +0x2bf
github.com/theupdateframework/go-tuf/v2/metadata/updater.(*Updater).onlineRefresh(0xc00025e600)
	/home/azeem/go/pkg/mod/github.com/theupdateframework/go-tuf/v2@v2.0.2/metadata/updater/updater.go:130 +0x45
github.com/theupdateframework/go-tuf/v2/metadata/updater.(*Updater).Refresh(0xc00025e600)
	/home/azeem/go/pkg/mod/github.com/theupdateframework/go-tuf/v2@v2.0.2/metadata/updater/updater.go:120 +0x98
github.com/sigstore/sigstore-go/pkg/tuf.(*Client).Refresh(0xc000362b58)
	/home/azeem/go/pkg/mod/github.com/sigstore/sigstore-go@v0.7.0/pkg/tuf/client.go:149 +0x158
github.com/sigstore/sigstore-go/pkg/tuf.(*Client).loadMetadata(0xc000362b58)
	/home/azeem/go/pkg/mod/github.com/sigstore/sigstore-go@v0.7.0/pkg/tuf/client.go:107 +0x22a
github.com/sigstore/sigstore-go/pkg/tuf.New(0xc0005a7e00)
	/home/azeem/go/pkg/mod/github.com/sigstore/sigstore-go@v0.7.0/pkg/tuf/client.go:84 +0x9c5
github.com/cli/cli/v2/pkg/cmd/attestation/verification.newPublicGoodVerifier()
	/home/azeem/Documents/dev/p/cli/pkg/cmd/attestation/verification/sigstore.go:294 +0x2a
github.com/cli/cli/v2/pkg/cmd/attestation/verification.(*LiveSigstoreVerifier).chooseVerifier(0xc0006c1d38, {0xc000162344, 0xc})
	/home/azeem/Documents/dev/p/cli/pkg/cmd/attestation/verification/sigstore.go:92 +0x17a
github.com/cli/cli/v2/pkg/cmd/attestation/verification.(*LiveSigstoreVerifier).verify(0xc0006dfd38, 0xc000352498, {0xc0005bb620?, {0xc000234088?, 0x2?, 0x2?}})
	/home/azeem/Documents/dev/p/cli/pkg/cmd/attestation/verification/sigstore.go:173 +0x14f
github.com/cli/cli/v2/pkg/cmd/attestation/verification.(*LiveSigstoreVerifier).Verify(0xc0006dfd38, {0xc000389ad0, 0x2, 0x15dd0a0?}, {0xc0005bb620?, {0xc000234088?, 0x2?, 0x2?}})
	/home/azeem/Documents/dev/p/cli/pkg/cmd/attestation/verification/sigstore.go:214 +0x245
github.com/cli/cli/v2/pkg/cmd/attestation/verification.TestLiveSigstoreVerifier(0xc0000e96c0)
	/home/azeem/Documents/dev/p/cli/pkg/cmd/attestation/verification/sigstore_integration_test.go:55 +0x516
testing.tRunner(0xc0000e96c0, 0xffd168)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/testing/testing.go:1690 +0x227
created by testing.(*T).Run in goroutine 1
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/testing/testing.go:1743 +0x826

goroutine 6241 [select]:
net/http.(*http2clientStream).writeRequest(0xc00041c000, 0xc0004da3c0, 0x0)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/h2_bundle.go:8673 +0x14ec
net/http.(*http2clientStream).doRequest(0xc00041c000, 0xc0004da3c0, 0x0)
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/h2_bundle.go:8551 +0xd2
created by net/http.(*http2ClientConn).roundTrip in goroutine 6319
	/home/azeem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.5.linux-amd64/src/net/http/h2_bundle.go:8456 +0x9a5
FAIL	github.com/cli/cli/v2/pkg/cmd/attestation/verification	600.034s
```

**UPDATE**

Running tests locally on **Windows 10** also reproduces this.
However, it's a bit random but overall percentage is pretty high i.e. ~80%.

Command:

```
go test -tags=integration .\pkg\cmd\attestation\... -count 2 -failfast
```

Errors:

```
--- FAIL: TestVerifyAttestations (1.13s)
    --- FAIL: TestVerifyAttestations/all_attestations_pass_verification (1.12s)
        attestation_integration_test.go:53:
                Error Trace:    C:/cli-trunk/pkg/cmd/attestation/verify/attestation_integration_test.go:53
                Error:          "[0xc000719230]" should have 2 item(s), but has 1
                Test:           TestVerifyAttestations/all_attestations_pass_verification
```

```
--- FAIL: TestLiveSigstoreVerifier (3.09s)
    sigstore_integration_test.go:62:
                Error Trace:    C:/cli-trunk/pkg/cmd/attestation/verification/sigstore_integration_test.go:62
                Error:          Not equal:
                                expected: 2
                                actual  : 1
                Test:           TestLiveSigstoreVerifier
                Messages:       test case: with valid artifact and JSON lines file containing multiple Sigstore bundles
```

```
--- FAIL: TestVerifyAttestations (3.72s)
    --- FAIL: TestVerifyAttestations/passes_verification_with_2/3_attestations_passing_Sigstore_verification (1.42s)
        attestation_integration_test.go:65:
                Error Trace:    C:/cli-trunk/pkg/cmd/attestation/verify/attestation_integration_test.go:65
                Error:          "[0xc000781d30]" should have 2 item(s), but has 1
                Test:           TestVerifyAttestations/passes_verification_with_2/3_attestations_passing_Sigstore_verification
```

```
--- FAIL: TestVerifyAttestations (1.85s)
    --- FAIL: TestVerifyAttestations/passes_verification_with_2/3_attestations_passing_Sigstore_verification (0.18s)
        attestation_integration_test.go:63:
                Error Trace:    C:/cli-trunk/pkg/cmd/attestation/verify/attestation_integration_test.go:63
                Error:          Received unexpected error:
                                failed to get bundle issuer: unsupported bundle version: application/vnd.dev.sigstore.bundle+json;version=0.1
                Test:           TestVerifyAttestations/passes_verification_with_2/3_attestations_passing_Sigstore_verification
```

```
--- FAIL: TestLiveSigstoreVerifier (3.61s)
    --- FAIL: TestLiveSigstoreVerifier/with_2/3_verified_attestations (0.82s)
        sigstore_integration_test.go:79:
                Error Trace:    C:/cli-trunk/pkg/cmd/attestation/verification/sigstore_integration_test.go:79
                Error:          "[0xc000293310]" should have 2 item(s), but has 1
                Test:           TestLiveSigstoreVerifier/with_2/3_verified_attestations
```

---

Following two JSON files are generated by running these tests:

1. `pkg/cmd/attestation/verification/tuf-repo.github.com.json`
2. `pkg/cmd/attestation/verification/tuf-repo-cdn.sigstore.dev.json`

and are not removed automatically.
Looks like they may be removed when the tests finish and/or added to `.gitignore`.

FWIW this [branch](https://github.com/cli/cli/commits/wm/narrow-flakes/) contains some modifications that I believe only run 2 tests, and it fails very frequently (after I bumped the specs on my Windows VM):

```
PS C:\Users\williammartin\workspace\cli> go test -v -tags=integration ./pkg/cmd/attestation/verification ./pkg/cmd/attestation/verify -count 2 -failfast
=== RUN   TestLiveSigstoreVerifier
=== RUN   TestLiveSigstoreVerifier/with_2/3_verified_attestations
--- PASS: TestLiveSigstoreVerifier (0.78s)
    --- PASS: TestLiveSigstoreVerifier/with_2/3_verified_attestations (0.78s)
=== RUN   TestLiveSigstoreVerifier
=== RUN   TestLiveSigstoreVerifier/with_2/3_verified_attestations
--- PASS: TestLiveSigstoreVerifier (0.59s)
    --- PASS: TestLiveSigstoreVerifier/with_2/3_verified_attestations (0.59s)
PASS
ok      github.com/cli/cli/v2/pkg/cmd/attestation/verification  1.499s
=== RUN   TestVerifyAttestations
=== RUN   TestVerifyAttestations/all_attestations_pass_verification
    attestation_integration_test.go:53:
                Error Trace:    C:/Users/williammartin/workspace/cli/pkg/cmd/attestation/verify/attestation_integration_test.go:53
                Error:          "[0x400045ae30]" should have 2 item(s), but has 1
                Test:           TestVerifyAttestations/all_attestations_pass_verification
--- FAIL: TestVerifyAttestations (0.43s)
    --- FAIL: TestVerifyAttestations/all_attestations_pass_verification (0.43s)
FAIL
FAIL    github.com/cli/cli/v2/pkg/cmd/attestation/verify        0.548s
FAIL
```

--- a/pkg/cmd/attestation/trustedroot/trustedroot.go
+++ b/pkg/cmd/attestation/trustedroot/trustedroot.go
@@ -11,6 +11,7 @@ import (
 	"github.com/cli/cli/v2/pkg/cmd/attestation/io"
 	"github.com/cli/cli/v2/pkg/cmd/attestation/verification"
 	"github.com/cli/cli/v2/pkg/cmdutil"
+	o "github.com/cli/cli/v2/pkg/option"
 	ghauth "github.com/cli/go-gh/v2/pkg/auth"
 
 	"github.com/MakeNowJust/heredoc"
@@ -121,7 +122,7 @@ func getTrustedRoot(makeTUF tufClientInstantiator, opts *Options) error {
 	var tufOptions []tufConfig
 	var defaultTR = "trusted_root.json"
 
-	tufOpt := verification.DefaultOptionsWithCacheSetting()
+	tufOpt := verification.DefaultOptionsWithCacheSetting(o.None[string]())
 	// Disable local caching, so we get up-to-date response from TUF repository
 	tufOpt.CacheValidity = 0
 
@@ -150,7 +151,7 @@ func getTrustedRoot(makeTUF tufClientInstantiator, opts *Options) error {
 			targets:    []string{defaultTR},
 		})
 
-		tufOpt = verification.GitHubTUFOptions()
+		tufOpt = verification.GitHubTUFOptions(o.None[string]())
 		tufOpt.CacheValidity = 0
 		tufOptions = append(tufOptions, tufConfig{
 			tufOptions: tufOpt,

--- a/pkg/cmd/attestation/verification/sigstore.go
+++ b/pkg/cmd/attestation/verification/sigstore.go
@@ -10,6 +10,7 @@ import (
 
 	"github.com/cli/cli/v2/pkg/cmd/attestation/api"
 	"github.com/cli/cli/v2/pkg/cmd/attestation/io"
+	o "github.com/cli/cli/v2/pkg/option"
 
 	"github.com/sigstore/sigstore-go/pkg/bundle"
 	"github.com/sigstore/sigstore-go/pkg/root"
@@ -34,6 +35,8 @@ type SigstoreConfig struct {
 	NoPublicGood bool
 	// If tenancy mode is not used, trust domain is empty
 	TrustDomain string
+	// TUFMetadataDir
+	TUFMetadataDir o.Option[string]
 }
 
 type SigstoreVerifier interface {
@@ -45,7 +48,8 @@ type LiveSigstoreVerifier struct {
 	Logger       *io.Handler
 	NoPublicGood bool
 	// If tenancy mode is not used, trust domain is empty
-	TrustDomain string
+	TrustDomain    string
+	TUFMetadataDir o.Option[string]
 }
 
 var ErrNoAttestationsVerified = errors.New("no attestations were verified")
@@ -55,10 +59,11 @@ var ErrNoAttestationsVerified = errors.New("no attestations were verified")
 // Public Good, GitHub, or a custom trusted root.
 func NewLiveSigstoreVerifier(config SigstoreConfig) *LiveSigstoreVerifier {
 	return &LiveSigstoreVerifier{
-		TrustedRoot:  config.TrustedRoot,
-		Logger:       config.Logger,
-		NoPublicGood: config.NoPublicGood,
-		TrustDomain:  config.TrustDomain,
+		TrustedRoot:    config.TrustedRoot,
+		Logger:         config.Logger,
+		NoPublicGood:   config.NoPublicGood,
+		TrustDomain:    config.TrustDomain,
+		TUFMetadataDir: config.TUFMetadataDir,
 	}
 }
 
@@ -89,9 +94,9 @@ func (v *LiveSigstoreVerifier) chooseVerifier(issuer string) (*verify.SignedEnti
 			if v.NoPublicGood {
 				return nil, fmt.Errorf("detected public good instance but requested verification without public good instance")
 			}
-			return newPublicGoodVerifier()
+			return newPublicGoodVerifier(v.TUFMetadataDir)
 		case GitHubIssuerOrg:
-			return newGitHubVerifier(v.TrustDomain)
+			return newGitHubVerifier(v.TrustDomain, v.TUFMetadataDir)
 		default:
 			return nil, fmt.Errorf("leaf certificate issuer is not recognized")
 		}
@@ -255,10 +260,10 @@ func newCustomVerifier(trustedRoot *root.TrustedRoot) (*verify.SignedEntityVerif
 	return gv, nil
 }
 
-func newGitHubVerifier(trustDomain string) (*verify.SignedEntityVerifier, error) {
+func newGitHubVerifier(trustDomain string, tufMetadataDir o.Option[string]) (*verify.SignedEntityVerifier, error) {
 	var tr string
 
-	opts := GitHubTUFOptions()
+	opts := GitHubTUFOptions(tufMetadataDir)
 	client, err := tuf.New(opts)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create TUF client: %v", err)
@@ -289,8 +294,8 @@ func newGitHubVerifierWithTrustedRoot(trustedRoot *root.TrustedRoot) (*verify.Si
 	return gv, nil
 }
 
-func newPublicGoodVerifier() (*verify.SignedEntityVerifier, error) {
-	opts := DefaultOptionsWithCacheSetting()
+func newPublicGoodVerifier(tufMetadataDir o.Option[string]) (*verify.SignedEntityVerifier, error) {
+	opts := DefaultOptionsWithCacheSetting(tufMetadataDir)
 	client, err := tuf.New(opts)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create TUF client: %v", err)

--- a/pkg/cmd/attestation/verification/sigstore_integration_test.go
+++ b/pkg/cmd/attestation/verification/sigstore_integration_test.go
@@ -9,6 +9,7 @@ import (
 	"github.com/cli/cli/v2/pkg/cmd/attestation/artifact"
 	"github.com/cli/cli/v2/pkg/cmd/attestation/io"
 	"github.com/cli/cli/v2/pkg/cmd/attestation/test"
+	o "github.com/cli/cli/v2/pkg/option"
 
 	"github.com/sigstore/sigstore-go/pkg/verify"
 	"github.com/stretchr/testify/require"
@@ -50,7 +51,8 @@ func TestLiveSigstoreVerifier(t *testing.T) {
 	for _, tc := range testcases {
 		t.Run(tc.name, func(t *testing.T) {
 			verifier := NewLiveSigstoreVerifier(SigstoreConfig{
-				Logger: io.NewTestHandler(),
+				Logger:         io.NewTestHandler(),
+				TUFMetadataDir: o.Some(t.TempDir()),
 			})
 
 			results, err := verifier.Verify(tc.attestations, publicGoodPolicy(t))
@@ -68,7 +70,8 @@ func TestLiveSigstoreVerifier(t *testing.T) {
 
 	t.Run("with 2/3 verified attestations", func(t *testing.T) {
 		verifier := NewLiveSigstoreVerifier(SigstoreConfig{
-			Logger: io.NewTestHandler(),
+			Logger:         io.NewTestHandler(),
+			TUFMetadataDir: o.Some(t.TempDir()),
 		})
 
 		invalidBundle := getAttestationsFor(t, "../test/data/sigstore-js-2.1.0-bundle-v0.1.json")
@@ -84,7 +87,8 @@ func TestLiveSigstoreVerifier(t *testing.T) {
 
 	t.Run("fail with 0/2 verified attestations", func(t *testing.T) {
 		verifier := NewLiveSigstoreVerifier(SigstoreConfig{
-			Logger: io.NewTestHandler(),
+			Logger:         io.NewTestHandler(),
+			TUFMetadataDir: o.Some(t.TempDir()),
 		})
 
 		invalidBundle := getAttestationsFor(t, "../test/data/sigstore-js-2.1.0-bundle-v0.1.json")
@@ -107,7 +111,8 @@ func TestLiveSigstoreVerifier(t *testing.T) {
 		attestations := getAttestationsFor(t, "../test/data/github_provenance_demo-0.0.12-py3-none-any-bundle.jsonl")
 
 		verifier := NewLiveSigstoreVerifier(SigstoreConfig{
-			Logger: io.NewTestHandler(),
+			Logger:         io.NewTestHandler(),
+			TUFMetadataDir: o.Some(t.TempDir()),
 		})
 
 		results, err := verifier.Verify(attestations, githubPolicy)
@@ -119,8 +124,9 @@ func TestLiveSigstoreVerifier(t *testing.T) {
 		attestations := getAttestationsFor(t, "../test/data/sigstore-js-2.1.0_with_2_bundles.jsonl")
 
 		verifier := NewLiveSigstoreVerifier(SigstoreConfig{
-			Logger:      io.NewTestHandler(),
-			TrustedRoot: test.NormalizeRelativePath("../test/data/trusted_root.json"),
+			Logger:         io.NewTestHandler(),
+			TrustedRoot:    test.NormalizeRelativePath("../test/data/trusted_root.json"),
+			TUFMetadataDir: o.Some(t.TempDir()),
 		})
 
 		results, err := verifier.Verify(attestations, publicGoodPolicy(t))

--- a/pkg/cmd/attestation/verification/tuf.go
+++ b/pkg/cmd/attestation/verification/tuf.go
@@ -5,6 +5,7 @@ import (
 	"os"
 	"path/filepath"
 
+	o "github.com/cli/cli/v2/pkg/option"
 	"github.com/cli/go-gh/v2/pkg/config"
 	"github.com/sigstore/sigstore-go/pkg/tuf"
 )
@@ -14,7 +15,7 @@ var githubRoot []byte
 
 const GitHubTUFMirror = "https://tuf-repo.github.com"
 
-func DefaultOptionsWithCacheSetting() *tuf.Options {
+func DefaultOptionsWithCacheSetting(tufMetadataDir o.Option[string]) *tuf.Options {
 	opts := tuf.DefaultOptions()
 
 	// The CODESPACES environment variable will be set to true in a Codespaces workspace
@@ -25,8 +26,8 @@ func DefaultOptionsWithCacheSetting() *tuf.Options {
 		opts.DisableLocalCache = true
 	}
 
-	// Set the cache path to a directory owned by the CLI
-	opts.CachePath = filepath.Join(config.CacheDir(), ".sigstore", "root")
+	// Set the cache path to the provided dir, or a directory owned by the CLI
+	opts.CachePath = tufMetadataDir.UnwrapOr(filepath.Join(config.CacheDir(), ".sigstore", "root"))
 
 	// Allow TUF cache for 1 day
 	opts.CacheValidity = 1
@@ -34,8 +35,8 @@ func DefaultOptionsWithCacheSetting() *tuf.Options {
 	return opts
 }
 
-func GitHubTUFOptions() *tuf.Options {
-	opts := DefaultOptionsWithCacheSetting()
+func GitHubTUFOptions(tufMetadataDir o.Option[string]) *tuf.Options {
+	opts := DefaultOptionsWithCacheSetting(tufMetadataDir)
 
 	opts.Root = githubRoot
 	opts.RepositoryBaseURL = GitHubTUFMirror

--- a/pkg/cmd/attestation/verification/tuf_test.go
+++ b/pkg/cmd/attestation/verification/tuf_test.go
@@ -5,16 +5,22 @@ import (
 	"path/filepath"
 	"testing"
 
+	o "github.com/cli/cli/v2/pkg/option"
 	"github.com/cli/go-gh/v2/pkg/config"
 	"github.com/stretchr/testify/require"
 )
 
-func TestGitHubTUFOptions(t *testing.T) {
+func TestGitHubTUFOptionsNoMetadataDir(t *testing.T) {
 	os.Setenv("CODESPACES", "true")
-	opts := GitHubTUFOptions()
+	opts := GitHubTUFOptions(o.None[string]())
 
 	require.Equal(t, GitHubTUFMirror, opts.RepositoryBaseURL)
 	require.NotNil(t, opts.Root)
 	require.True(t, opts.DisableLocalCache)
 	require.Equal(t, filepath.Join(config.CacheDir(), ".sigstore", "root"), opts.CachePath)
+}
+
+func TestGitHubTUFOptionsWithMetadataDir(t *testing.T) {
+	opts := GitHubTUFOptions(o.Some("anything"))
+	require.Equal(t, "anything", opts.CachePath)
 }

--- a/pkg/cmd/attestation/verify/attestation_integration_test.go
+++ b/pkg/cmd/attestation/verify/attestation_integration_test.go
@@ -10,6 +10,7 @@ import (
 	"github.com/cli/cli/v2/pkg/cmd/attestation/io"
 	"github.com/cli/cli/v2/pkg/cmd/attestation/test"
 	"github.com/cli/cli/v2/pkg/cmd/attestation/verification"
+	o "github.com/cli/cli/v2/pkg/option"
 	"github.com/sigstore/sigstore-go/pkg/fulcio/certificate"
 	"github.com/stretchr/testify/require"
 )
@@ -25,7 +26,8 @@ func getAttestationsFor(t *testing.T, bundlePath string) []*api.Attestation {
 
 func TestVerifyAttestations(t *testing.T) {
 	sgVerifier := verification.NewLiveSigstoreVerifier(verification.SigstoreConfig{
-		Logger: io.NewTestHandler(),
+		Logger:         io.NewTestHandler(),
+		TUFMetadataDir: o.Some(t.TempDir()),
 	})
 
 	certSummary := certificate.Summary{}

--- a/pkg/cmd/attestation/verify/verify_integration_test.go
+++ b/pkg/cmd/attestation/verify/verify_integration_test.go
@@ -11,6 +11,7 @@ import (
 	"github.com/cli/cli/v2/pkg/cmd/attestation/test"
 	"github.com/cli/cli/v2/pkg/cmd/attestation/verification"
 	"github.com/cli/cli/v2/pkg/cmd/factory"
+	o "github.com/cli/cli/v2/pkg/option"
 	"github.com/cli/go-gh/v2/pkg/auth"
 	"github.com/stretchr/testify/require"
 )
@@ -19,7 +20,8 @@ func TestVerifyIntegration(t *testing.T) {
 	logger := io.NewTestHandler()
 
 	sigstoreConfig := verification.SigstoreConfig{
-		Logger: logger,
+		Logger:         logger,
+		TUFMetadataDir: o.Some(t.TempDir()),
 	}
 
 	cmdFactory := factory.New("test")
@@ -130,7 +132,8 @@ func TestVerifyIntegrationCustomIssuer(t *testing.T) {
 	logger := io.NewTestHandler()
 
 	sigstoreConfig := verification.SigstoreConfig{
-		Logger: logger,
+		Logger:         logger,
+		TUFMetadataDir: o.Some(t.TempDir()),
 	}
 
 	cmdFactory := factory.New("test")
@@ -200,7 +203,8 @@ func TestVerifyIntegrationReusableWorkflow(t *testing.T) {
 	logger := io.NewTestHandler()
 
 	sigstoreConfig := verification.SigstoreConfig{
-		Logger: logger,
+		Logger:         logger,
+		TUFMetadataDir: o.Some(t.TempDir()),
 	}
 
 	cmdFactory := factory.New("test")
@@ -289,7 +293,8 @@ func TestVerifyIntegrationReusableWorkflowSignerWorkflow(t *testing.T) {
 	logger := io.NewTestHandler()
 
 	sigstoreConfig := verification.SigstoreConfig{
-		Logger: logger,
+		Logger:         logger,
+		TUFMetadataDir: o.Some(t.TempDir()),
 	}
 
 	cmdFactory := factory.New("test")

